### PR TITLE
Create form for create and update post

### PR DESCRIPTION
**Pull request by serhii73**

### CODE REVIEW

**Comment by serhii73** [comment]
@webknjaz we can merge this Django project in master?

**Comment by webknjaz** [comment]
No

File: blog/forms.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""Forms for the blog app."""
+from django import forms
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Create and edit form."""
+
+    class Meta:
+        """Build form on the model and with the fields."""
```
Comment (webknjaz) [inline blog/forms.py]: It's not an action.

File: blog/forms.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+"""Forms for the blog app."""
+from django import forms
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Create and edit form."""
```
Comment (webknjaz) [inline blog/forms.py]: It's not an action, but entity.

File: blog/templates/blog/base.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{% load staticfiles %}
+<html>
+<head>
+    <title>Django Girls blog</title>
+    <link rel="stylesheet"
+          href="//maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css">
+    <link rel="stylesheet"
+          href="//maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap-theme.min.css">
+    <link href='//fonts.googleapis.com/css?family=Lobster&subset=latin,latin-ext'
```
Comment (webknjaz) [inline blog/templates/blog/base.html]: Don't mix different quotation styles.

File: blog/templates/blog/post_detail.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+{% block content %}
+    <div class="post">
+        {% if post.published_date %}
+            <div class="date">
+                {{ post.published_date }}
+            </div>
+        {% endif %}
+        {% if user.is_authenticated %}
+            <a class="btn btn-default"
+               href="{% url 'post_edit' pk=post.pk %}"><span
```
Comment (webknjaz) [inline blog/templates/blog/post_detail.html]: Reformat this with the hierarchy in mind, not just do a new line on random space.

File: blog/forms.py (hunk 2 above)
Comment (serhii73) [reply blog/forms.py]: @webknjaz `Creating and editing forms.` ?  Is good doc for this `class`?

File: blog/templates/blog/post_detail.html (hunk 4 above)
Comment (serhii73) [reply blog/templates/blog/post_detail.html]: @webknjaz Sorry?

**Comment by serhii73** [comment]
@webknjaz I fixed all problems.

File: blog/templates/blog/post_list.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{% extends "blog/base.html" %}
+
+{% block content %}
+    {% for post in posts %}
+        <div class="post">
+            <div class="date">
+                <p>Опубликовано: {{ post.published_date }}</p>
```
Comment (webknjaz) [inline blog/templates/blog/post_list.html]: Please replace **all** translatable strings with i18n (`{% trans "Posted on:" %}`).
This requires `{% load i18n %}`.

File: blog/templates/blog/post_detail.html [hunk 6]
```diff
@@ -0,0 +1,18 @@
+{% extends "blog/base.html" %}
+
+{% block content %}
+    <div class="post">
+        {% if user.is_authenticated %}
+            <a class="btn btn-default"
+               href="{% url 'post_edit' pk=post.pk %}"><span
+                    class="glyphicon glyphicon-pencil"></span></a>
```
Comment (webknjaz) [inline blog/templates/blog/post_detail.html]: make this nicer

File: blog/templates/blog/base.html [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+          href="//maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap-theme.min.css">
+    <link href="//fonts.googleapis.com/css?family=Lobster&subset=latin,latin-ext"
+          rel="stylesheet" type="text/css">
+    <link rel="stylesheet" href="{% static 'css/blog.css' %}">
+</head>
+<body>
+<div class="page-header">
+    {% if user.is_authenticated %}
+        <a href="{% url 'post_new' %}" class="top-menu"><span
+                class="glyphicon glyphicon-plus"></span></a>
```
Comment (webknjaz) [inline blog/templates/blog/base.html]: this kind of nesting is unreadable

**Comment by serhii73** [comment]
@webknjaz Sorry, 

> pyflakes - This hook runs pyflakes. (This is deprecated, use flake8).

from [pre-commit.com/hooks.html](https://pre-commit.com/hooks.html)
If I right to understand it says here that I need to use `flake8` and [my code has it.](https://github.com/kpi-web-guild/django-girls-blog-serhii73/blob/feature/6/django_forms/.pre-commit-config.yaml#L11)

**Comment by serhii73** [comment]
@webknjaz 

> It uses only a limited subset of pyflakes. try pylint then

I need to create a new branch or add `pylint` to pre-commit here?

**Comment by webknjaz** [comment]
Let's do it here

**Comment by serhii73** [comment]
```
************* Module blog.admin
E:  4, 0: Unable to import 'django.contrib' (import-error)
************* Module blog.apps
E:  2, 0: Unable to import 'django.apps' (import-error)
R:  5, 0: Too few public methods (0/2) (too-few-public-methods)
************* Module blog.forms
E:  2, 0: Unable to import 'django' (import-error)
R:  9, 4: Too few public methods (0/2) (too-few-public-methods)
R:  6, 0: Too few public methods (0/2) (too-few-public-methods)
************* Module blog.migrations.0001_initial
C: 30, 0: Wrong continued indentation (remove 3 spaces).
                    models.DateTimeField(blank=True, null=True)),
                 |  ^ (bad-continuation)
C:  1, 0: Module name "0001_initial" doesn't conform to snake_case naming style (invalid-name)
C:  1, 0: Missing module docstring (missing-docstring)
E:  3, 0: Unable to import 'django.conf' (import-error)
E:  4, 0: Unable to import 'django.db' (import-error)
E:  5, 0: Unable to import 'django.db.models.deletion' (import-error)
E:  6, 0: Unable to import 'django.utils.timezone' (import-error)
C:  9, 0: Missing class docstring (missing-docstring)
R:  9, 0: Too few public methods (0/2) (too-few-public-methods)
************* Module blog.models
C: 13, 0: Wrong hanging indentation (remove 4 spaces).
            default=timezone.now)
        |   ^ (bad-continuation)
C: 15, 0: Wrong hanging indentation (remove 4 spaces).
            blank=True, null=True)
        |   ^ (bad-continuation)
E:  2, 0: Unable to import 'django.db' (import-error)
E:  3, 0: Unable to import 'django.utils' (import-error)
************* Module blog.urls
E:  2, 0: Unable to import 'django.urls' (import-error)
C:  6, 0: Constant name "urlpatterns" doesn't conform to UPPER_CASE naming style (invalid-name)
************* Module blog.views
E:  2, 0: Unable to import 'django.shortcuts' (import-error)
E:  3, 0: Unable to import 'django.utils' (import-error)
C: 19, 0: Argument name "pk" doesn't conform to snake_case naming style (invalid-name)
C: 40, 0: Argument name "pk" doesn't conform to snake_case naming style (invalid-name)
************* Module mysite.urls
E:  2, 0: Unable to import 'django.contrib' (import-error)
E:  3, 0: Unable to import 'django.urls' (import-error)
C:  5, 0: Constant name "urlpatterns" doesn't conform to UPPER_CASE naming style (invalid-name)
************* Module mysite.wsgi
E: 12, 0: Unable to import 'django.core.wsgi' (import-error)
C: 16, 0: Constant name "application" doesn't conform to UPPER_CASE naming style (invalid-name)
-----------------------------------
Your code has been rated at 1.35/10
No config file found, using default configuration
```
https://github.com/pre-commit/pre-commit-hooks/issues/157 - Do I need to use a local hook? - https://github.com/Yelp/venv-update/blob/de9e1befdecf448cf02ceb5c57a30bcc88a58c06/.pre-commit-config.yaml#L30-L36
Or maybe to add other hooks instead pylint? - https://pre-commit.com/hooks.html

**Comment by serhii73** [comment]
@webknjaz Hi. What do you say about the code? Good now or still bad?
And what I need to do with `pylint`? Maybe add instead another checking instrument?
Because pylint not good working.

**Comment by webknjaz** [comment]
It's not pylint being broken, it's you misconfiguring it.

File: blog/templates/blog/base.html [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{% load staticfiles %}
+
+{% load i18n %}
+
+<html>
+<head>
+    <title>Django Girls blog</title>
```
Comment (webknjaz) [inline blog/templates/blog/base.html]: i18n

**Comment by webknjaz** [comment]
I've done some research and put together a config for you:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index adec32c..a22d1d8 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -18,3 +18,23 @@
   hooks:
   - id: pydocstyle
     exclude: blog/migrations
+
+- repo: git://github.com/pre-commit/mirrors-pylint.git
+  rev: v1.9.1
+  hooks:
+  - id: pylint
+    args:
+    - --msg-template={path}:L {line:3d}:({symbol}) {msg} (C {column:2d}), ::{msg_id}
+    - --output-format=colorized
+    - |-
+        --init-hook=
+        import os, platform, sys;
+        sys.path.insert(
+          0,
+          "".join([os.getenv("HOME"), "/virtualenv/python", platform.python_version(), "/site-packages"])
+          if os.getenv("TRAVIS") else
+          "{}/test_venv/lib/python3.6/site-packages".format(os.getcwd())
+        )  # inject virtualenv packages into a module search path; could also go for pylint-venv hook
+    - --load-plugins=pylint_django
+    - --good-names=pk,urlpatterns
+    exclude: blog/migrations/
```

(you'll need to pre-install `pylint-django` plugin though)

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -18,3 +18,10 @@
   hooks:
   - id: pydocstyle
     exclude: blog/migrations
+
+- repo: https://github.com/pre-commit/mirrors-pylint
+  rev: v1.9.0
+  hooks:
+  - id: pylint
+    exclude: blog/migrations
+    args : [--rcfile=.pylintrc]
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: drop this file and use config I posted

File: .pylintrc [hunk 10]
```diff
@@ -0,0 +1,548 @@
+[MASTER]
```
Comment (webknjaz) [inline .pylintrc]: Remove this config.

File: blog/forms.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Class form based on the model."""
+
+    class Meta:
+        """Choose a model for creation form.
```
Comment (webknjaz) [inline blog/forms.py]: Does linter really require docstrings for in-body classes?

Also, it's not an action, but metadata specification/description.

**Comment by serhii73** [comment]
@webknjaz 
```
hookid: pydocstyle

blog/forms.py:10 in public nested class `Meta`:
        D106: Missing docstring in public nested class
```
it the answer for the question 

> Does linter really require docstrings for in-body classes?

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -18,3 +18,23 @@
...
+    - --msg-template={path}:L {line:3d}:({symbol}) {msg} (C {column:2d}), ::{msg_id}
+    - --output-format=colorized
+    - |-
+        --init-hook=
+        import os, platform, sys;
+        sys.path.insert(
+          0,
+          "".join([os.getenv("HOME"), "/virtualenv/python", platform.python_version(), "/site-packages"])
+          if os.getenv("TRAVIS") else
+          "{}/test_venv/lib/python3.6/site-packages".format(os.getcwd())
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: Do you use `test_venv` in your local machine for a virtual env or not? If it's something else, be sure to change this folder name.

File: .travis.yml [hunk 13]
```diff
@@ -2,4 +2,5 @@ language: python
 python: 3.6
 cache: pip
 before_install: pip install -U pre-commit
+before_install: pip install pylint-django==0.11.1
```
Comment (webknjaz) [inline .travis.yml]: It's an invalid YAML because of key duplication.

**Comment by webknjaz** [comment]
Could you please reorder commits so that adding linting would go first?

File: .travis.yml [hunk 14]
```diff
@@ -1,6 +1,7 @@
 language: python
 python: 3.6
 cache: pip
-before_install: pip install -U pre-commit
-before_install: pip install pylint-django==0.11.1
+before_install:
+- pip install -U pre-commit
+- pip install pylint-django==0.11.1
```
Comment (webknjaz) [inline .travis.yml]: you could install them with a single command, but it's okay for now.

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -18,3 +18,24 @@
...
+    args:
+    - --msg-template={path}:L {line:3d}:({symbol}) {msg} (C {column:2d}), ::{msg_id}
+    - --output-format=colorized
+    - |-
+        --init-hook=
+        import os, platform, sys;
+        sys.path.insert(
+          0,
+          "".join([os.getenv("HOME"), "/virtualenv/python", platform.python_version(),
+           "/lib/python", ".".join(platform.python_version_tuple()[:2]), "/site-packages"])
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: make indentation nicer.

File: blog/models.py [hunk 16]
```diff
@@ -10,9 +10,9 @@ class Post(models.Model):
     title = models.CharField(max_length=200)
     text = models.TextField()
     created_date = models.DateTimeField(
-            default=timezone.now)
+        default=timezone.now)
```
Comment (webknjaz) [inline blog/models.py]: This fits in one line.

File: blog/models.py [hunk 17]
```diff
@@ -10,9 +10,9 @@ class Post(models.Model):
...
     published_date = models.DateTimeField(
-            blank=True, null=True)
+        blank=True, null=True)
```
Comment (webknjaz) [inline blog/models.py]: as well as this.

File: mysite/urls.py [hunk 18]
```diff
@@ -1,7 +1,8 @@
 """mysite URL Configuration."""
 from django.contrib import admin
-from django.urls import path
+from django.urls import include, path
 
 urlpatterns = [
     path('admin/', admin.site.urls),
+    path(r'', include('blog.urls')),
```
Comment (webknjaz) [inline mysite/urls.py]: I don't think you need a raw-string literal here.

File: requirements.txt [hunk 19]
```diff
@@ -0,0 +1,4 @@
+Django==2.0.5
```
Comment (webknjaz) [inline requirements.txt]: :+1:

File: requirements.txt [hunk 20]
```diff
@@ -0,0 +1,4 @@
...
+pre-commit==1.10.1
+pylint==1.9.2
```
Comment (webknjaz) [inline requirements.txt]: :-1: you don't need this in the current env, only in pre-commit-created one.

File: .travis.yml [hunk 21]
```diff
@@ -1,5 +1,8 @@
 language: python
 python: 3.6
 cache: pip
-before_install: pip install -U pre-commit
+before_install:
+- pip install -U pre-commit
+- pip install pylint-django==0.11.1
+- pip install -r requirements.txt
```
Comment (webknjaz) [inline .travis.yml]: Why did you split just one requirement into a file, while listing others outside of it? Be consistent.

**Comment by webknjaz** [comment]
clean-up your commits

**Comment by serhii73** [comment]
@webknjaz Now is good?

File: .pre-commit-config.yaml [hunk 22]
```diff
@@ -18,3 +18,24 @@
...
+    - --output-format=colorized
+    - |-
+        --init-hook=
+        import os, platform, sys;
+        sys.path.insert(
+        0,
+        "".join([os.getenv("HOME"), "/virtualenv/python", platform.python_version(),
+        "/lib/python", ".".join(platform.python_version_tuple()[:2]), "/site-packages"])
+        if os.getenv("TRAVIS") else
+        "/home/sergey/.virtualenv/django_girls/lib/python3.6/site-packages".format(os.getcwd())
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: **NEVER** hardcode paths. Also, you inject current working directory from `os.getcwd()` to nowhere.

File: blog/templates/blog/post_edit.html [hunk 23]
```diff
@@ -0,0 +1,13 @@
+{% extends "blog/base.html" %}
+
+{% load i18n %}
+
+{% block content %}
+    <h1>New post</h1>
```
Comment (webknjaz) [inline blog/templates/blog/post_edit.html]: i18n.

File: blog/views.py [hunk 24]
```diff
@@ -1,3 +1,53 @@
 """Views for the blog app."""
+from django.shortcuts import get_object_or_404, redirect, render
+from django.utils import timezone
 
-# Create your views here.
+from .forms import PostForm
+from .models import Post
+
+
+def post_list(request):
```
Comment (webknjaz) [inline blog/views.py]: This is unrelated to the PR purpose. You need to merge #25/#24 first.
To simplify things, move pylint integration commit into a separate PR and let's merge it before other things.

**Comment by serhii73** [comment]
@webknjaz Done.

File: blog/forms.py [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Class form based on the model."""
+
+    class Meta:
+        """Description metadata."""
```
Comment (webknjaz) [inline blog/forms.py]: which description?

File: blog/forms.py [hunk 26]
```diff
@@ -0,0 +1,14 @@
+"""Creating and editing forms."""
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Class form based on the model."""
```
Comment (webknjaz) [inline blog/forms.py]: i see that it's a class, it's stated one line above. anything useful you want to tell me with this?

File: blog/forms.py [hunk 27]
```diff
@@ -0,0 +1,14 @@
+"""Creating and editing forms."""
```
Comment (webknjaz) [inline blog/forms.py]: Is it a prose or a module description? It's not an action, it's a module.

File: mysite/urls.py [hunk 28]
```diff
@@ -4,5 +4,5 @@
 
 urlpatterns = [
     path('admin/', admin.site.urls),
-    path(r'', include('blog.urls')),
+    path('', include('blog.urls')),
```
Comment (webknjaz) [inline mysite/urls.py]: How is this change related to this PR?

**Comment by serhii73** [comment]
@webknjaz what do you think now about the code?

**Comment by webknjaz** [comment]
I think that you need to read all comments, not just ones you like.

File: blog/forms.py [hunk 29]
```diff
@@ -0,0 +1,14 @@
+"""The module has forms who saved in DB and will be shown in HTML file."""
```
Comment (webknjaz) [inline blog/forms.py]: Mind your grammar

File: blog/forms.py [hunk 30]
```diff
@@ -0,0 +1,14 @@
...
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Form for our Post model."""
```
Comment (webknjaz) [inline blog/forms.py]: "our"? whose?

File: blog/forms.py [hunk 31]
```diff
@@ -0,0 +1,14 @@
...
+
+    class Meta:
+        """Description Post's metadata."""
```
Comment (webknjaz) [inline blog/forms.py]: Docstring is a description, it's obvious.

File: blog/forms.py [hunk 32]
```diff
@@ -0,0 +1,14 @@
...
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Form for Post model."""
+
+    class Meta:
+        """Configuration for form."""
```
Comment (webknjaz) [inline blog/forms.py]: `of`

File: blog/forms.py [hunk 33]
```diff
@@ -0,0 +1,14 @@
+"""The module has forms who saved in DB and will be shown in HTML file."""
+from django import forms
+
+from .models import Post
+
+
+class PostForm(forms.ModelForm):
+    """Form for Post model."""
```
Comment (webknjaz) [inline blog/forms.py]: Forget about model, it's about visual effect.
